chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+
+    The default trainer is `plain`, i.e., it does not contain any extensions.
+
+    Args:
+        updater (Updater): Updater object. It defines how to update the models.
+        stop_trigger: Trigger that determines when to stop the training loop.
+            If it is not callable, it is passed to :class:`IntervalTrigger`.
+
+    Attributes:
+        updater: The updater object for this trainer.
+        stop_trigger: Trigger that determines when to stop the training loop.
+            The training loop stops at the iteration on which this trigger
+            returns True.
+        observation: Observation of values made at the last update. See the
+            :class:`Reporter` class for details.
+        out: Output directory.
+        reporter: Reporter object to report observed values.
+
+    """
+    def __init__(self, updater, stop_trigger=None, out='result'):
+        self.updater = updater
+        self.stop_trigger = trigger_module.get_trigger(stop_trigger)
+        self.observation = {}
+        self.out = out
+
+        reporter = reporter_module.Reporter()
+        for name, optimizer in six.iteritems(updater.get_all_optimizers()):
+            reporter.add_observer(name, optimizer.target)
+            reporter.add_observers(
+                name, optimizer.target.namedlinks(skipself=True))
+        self.reporter = reporter
+
+        self._done = False
+        self._extensions = collections.OrderedDict()
+
+        updater.connect_trainer(self)
+
+    def extend(self, extension, name=None, trigger=None, priority=None,
+               invoke_before_training=None):
+        """Registers an extension to the trainer.
+
+        :class:`Extension` is a callable object which is called after each
+        update unless the corresponding trigger object decides to skip the
+        iteration. The order of execution is determined by priorities:
+        extensions with higher priorities are called earlier in each iteration.
+        Extensions with the same priority are invoked in the order of
+        registrations.
+
+        If two or more extensions with the same name are registered, suffixes
+        are added to the names of the second to last extensions. The suffix is
+        ``_N`` where N is the ordinal of the extensions.
+
+        See :class:`Extension` for the interface of extensions.
+
+        Args:
+            extension: Extension to register.
+            trigger (tuple or Trigger): Trigger object that determines when to
+                invoke the extension. If it is None, ``extension.trigger`` is
+                used instead. If the trigger is not callable, it is passed to
+                :class:`IntervalTrigger` to build an interval trigger.
+            priority (int): Invocation priority of the extension. Extensions
+                are invoked in the descending order of priorities in each
+                iteration. If this is None, ``extension.priority`` is used
```

Comment:
"If `extension` does not have `priority` attribute, the priority is set to PRIORITY_READER".